### PR TITLE
Add hook for data seralisation

### PR DESCRIPTION
**Pull request by rhyshort**
## What
Make it possible for operations to complete the operation with error when data serialisation fails

## How
Added new `serialise` hook which is a throwing function which allows operations to error when attempting to transform the data to json.

### CODE REVIEW

File: Source/CouchOperation.swift [hunk 1]
```diff
@@ -198,23 +198,42 @@ public class CouchOperation : NSOperation, HTTPRequestOperation
...
-            isFinished = true
-            return
-        }
+    /**
+     This should be used to serialise any data into the format expected by the CouchDB/Cloudant 
+     endpoint.
+     
+     - throws: An error in the event of a failure to serialise.
+     - note: This is guranteed to be  called after `validate() -> Bool` and before any of the 
+     properties defined in `HttpOperation` are called.
```
Comment (ricellis) [inline Source/CouchOperation.swift]: > before any of the properties defined in `HttpOperation` are called.

I think this is a bit confusing. Does it really mean before any of the property values are fixed in the request? Admittedly that isn't a much better description.

File: Tests/QueryViewTests.swift [hunk 2]
```diff
@@ -72,6 +72,8 @@ public class QueryViewTests : XCTestCase {
         view.stale = .Ok
         view.startKey = "startKey"
         view.databaseName = self.dbName
+        XCTAssert(view.validate())
+        try view.serialise()
         
         XCTAssert(view.validate())
```
Comment (ricellis) [inline Tests/QueryViewTests.swift]: This is now a duplicate call of `XCTAssert(view.validate())` (for all 3 tests you modified).

Comment (rhyshort) [inline Tests/QueryViewTests.swift]: fixed in 9f6a99a

File: Source/CouchOperation.swift (hunk 1 above)
Comment (rhyshort) [inline Source/CouchOperation.swift]: done in 1d94831

File: Source/CouchOperation.swift [hunk 3]
```diff
@@ -203,8 +203,8 @@ public class CouchOperation : NSOperation, HTTPRequestOperation
      endpoint.
      
      - throws: An error in the event of a failure to serialise.
-     - note: This is guranteed to be  called after `validate() -> Bool` and before any of the 
-     properties defined in `HttpOperation` are called.
+     - note: This is guranteed to be  called after `validate() -> Bool` and before the HttpOperation
```
Comment (ricellis) [inline Source/CouchOperation.swift]: was it intentional to remove the backticks from `HttpOperation`?

Comment (rhyshort) [inline Source/CouchOperation.swift]: fixed in b766c1d

**Comment by ricellis** [comment]
+1

**Pull request merged**